### PR TITLE
fix: Allow Singer schemas to include the `required` and `enum` fields

### PR DESCRIPTION
**Pull request by Jack-Burnett**
Closes [#901](https://github.com/meltano/sdk/issues/901)

The use of the Schema class from a dependency was limiting schemas to only include certain declared json schema fields, since it would remove properties if they were not whitelisted.  Since this is from a dependency, we can't simply expand the whitelist so instead we override the class and expand the list there. An alternative implementation would be to simply replace the class outright (no inheritance) - this could break existing taps that pass a Schema to Stream.init though.

It helps with but does not close [#332](https://github.com/meltano/sdk/issues/332), since 332 is asking for helpers, which this does not add - it makes it possible to add them though.

I think there are lots more improvements that could be made to SchemaPlus in follow-up PRs (adding many more legitimate properties to the whitelist and having better support for some non-simple properties.

Also, I removed the 'selected' and 'inclusion' fields from the new class - because they are part of the singer spec and not actually valid fields to see on a json schema.

### CODE REVIEW

**Comment by edgarrmondragon** [comment]
Hi @Jack-Burnett! Thanks for the PR. I'll take a look in the next few days 😺

File: singer_sdk/helpers/_schema.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        self.anyOf = anyOf
+        self.format = format
+        self.additionalProperties = additionalProperties
+        self.patternProperties = patternProperties
+        self.required = required
+        self.enum = enum
+        self.title = title
+
+    def __str__(self):
+        return json.dumps(self.to_dict())
```
Comment (edgarrmondragon) [inline singer_sdk/helpers/_schema.py:87]: This is implemented in the parent class, so you safely can omit this

```suggestion
```

File: singer_sdk/helpers/_schema.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    "enum",
+    # These are NOT simple keys (they can contain schemas themselves). We could
+    # consider adding extra handling to them.
+    "additionalProperties",
+    "anyOf",
+    "patternProperties",
+]
+
+
+class SchemaPlus(Schema):  # pylint: disable=too-many-instance-attributes
```
Comment (edgarrmondragon) [inline singer_sdk/helpers/_schema.py:32]: @Jack-Burnett wdyt of making this a `dataclass`, similar to `singer_sdk.helpers._singer.CatalogEntry`?

That'd would save some redundant code in the constructor and give us a nice `__repr__` and `__eq__` for free.

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]

File: singer_sdk/helpers/_schema.py (hunk 2 above)
Comment (Jack-Burnett) [reply singer_sdk/helpers/_schema.py:34]: Yeah that sounds like a good idea, done.
It also meant I had to give some type hints which are nice too, though I left a couple at Any for the more flexible json-schema fields

File: singer_sdk/helpers/_schema.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    # These are NOT simple keys (they can contain schemas themselves). We could
+    # consider adding extra handling to them.
+    "additionalProperties",
+    "anyOf",
+    "patternProperties",
+]
+
+
+@dataclass
+class SchemaPlus(Schema):  # pylint: disable=too-many-instance-attributes
```
Comment (edgarrmondragon) [inline singer_sdk/helpers/_schema.py:34]:
```suggestion
class SchemaPlus(Schema):
```

File: singer_sdk/helpers/_schema.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        result = {}
+
+        if self.properties is not None:
+            result["properties"] = {
+                k: v.to_dict()
+                for k, v in self.properties.items()  # pylint: disable=no-member
+            }
+
+        if self.items is not None:
+            result["items"] = self.items.to_dict()  # pylint: disable=no-member
```
Comment (edgarrmondragon) [inline singer_sdk/helpers/_schema.py:76]:
```suggestion
            result["items"] = self.items.to_dict()
```

**Comment by Jack-Burnett** [comment]
> @Jack-Burnett this is looking great!
> 
> I initially missed those pylint comments which we don't need. Once that's addressed this is g2g 🙂

Ohh yeah just lazy copy-pasting on my part 😅 All gone now

**Pull request merged**